### PR TITLE
Add gradle check jenkins workflow on public jenkins

### LINKED ISSUE #851: Convert freestyle gradle check to Jenkinsfile and migrate to public jenkins
### Is your feature request related to a problem? Please describe

Convert freestyle gradle check to Jenkinsfile.

### Describe the solution you'd like

https://github.com/opensearch-project/opensearch-build/issues/74#issuecomment-956496829

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by peterzhuamazon**
Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

### Description
Add gradle check jenkins workflow on public jenkins

### Issues Resolved
#851 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: vars/runGradleCheck.groovy [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                set +x
+
+                env | grep JAVA | grep HOME
+
+                echo "Git clone: ${git_repo_url} with ref: ${git_reference}"
+                rm -rf search
+                git clone ${git_repo_url} search
+                cd search/
+                git checkout -f ${git_reference}
+                git rev-parse HEAD
```
Comment (gaiksaya) [inline vars/runGradleCheck.groovy:31]: Why not use git groovy scripts here? Example https://github.com/opensearch-project/opensearch-build/blob/09f65b34a719581a40397a58b081b4dafc12ce5e/jenkins/opensearch-ruby/Jenkinsfile#L30

File: vars/runGradleCheck.groovy [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+                echo "Stop existing gradledaemon"
+                ./gradlew --stop
+                find ~/.gradle -type f -name "*.lock" -delete
+
+                echo "Check existing dockercontainer"
+                docker ps -a
+                docker stop `docker ps -qa` > /dev/null 2>&1 || echo
+                docker rm `docker ps -qa` > /dev/null 2>&1 || echo
+                echo "Stop existing dockercontainer"
```
Comment (gaiksaya) [inline vars/runGradleCheck.groovy:41]: Is this a precautionary measure? AFAIK each executor will have its own container right?

File: vars/runGradleCheck.groovy [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (git_repo_url.equals('null') || git_reference.equals('null')) {
+        println("No git repo url or git reference to checkout the commit, exit 1")
+        System.exit(1)
+    }
+    else {
+        withCredentials([
+            usernamePassword(credentialsId: "jenkins-gradle-check-s3-aws-credentials", usernameVariable: 'amazon_s3_access_key', passwordVariable: 'amazon_s3_secret_key'),
+            usernamePassword(credentialsId: "jenkins-gradle-check-s3-aws-resources", usernameVariable: 'amazon_s3_base_path', passwordVariable: 'amazon_s3_bucket'),
+            usernamePassword(credentialsId: "jenkins-github-bot-token", usernameVariable: 'GITHUB_BOT_USER', passwordVariable: 'GITHUB_BOT_TOKEN')]) {
```
Comment (gaiksaya) [inline vars/runGradleCheck.groovy:17]: Where are these used?

File: vars/runGradleCheck.groovy (hunk 1 above)
Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:31]: I have more control over here.
It is just a preference as majority of this is shell script.
So no need to use a jenkins specific groovy lib to do so.

File: vars/runGradleCheck.groovy (hunk 2 above)
Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:41]: Gradle check have issues with running on containers and has been confirmed with core engineers, it also does not help to run in containers as gradle check will spawn new containers for testing, thus cause a container in container situation which is hard to clean up.

I have synced with @bbarani and we are going to run directly on Agent node, just like fork jenkins setup now.

Therefore it is needed to make sure there is no existing docker container running as gradle check will spin up new containers.

File: vars/runGradleCheck.groovy (hunk 3 above)
Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:17]: These are used for connecting to an existing s3 bucket and test repository-s3 plugin functionality.
More details you can sync up with @dreamer-89 on it.
He created those credentials and demand the addition.

Thanks.

**Comment by dreamer-89** [comment]
test failure
```
TestUploadToS3 > testUploadToS3 STANDARD_OUT
    Loading shared library jenkins with version <notNeeded>
    Executing on agent [label:none]
       UploadToS3_Jenkinsfile.run()
          UploadToS3_Jenkinsfile.pipeline(groovy.lang.Closure)
             UploadToS3_Jenkinsfile.echo(Executing on agent [label:none])
             UploadToS3_Jenkinsfile.stage(sign, groovy.lang.Closure)
                UploadToS3_Jenkinsfile.script(groovy.lang.Closure)
                   UploadToS3_Jenkinsfile.uploadToS3({sourcePath=/tmp/src/path, bucket=dummy_bucket, path=/upload/path})
                      uploadToS3.string({credentialsId=jenkins-aws-account-public, variable=AWS_ACCOUNT_PUBLIC})
                      uploadToS3.withCredentials([AWS_ACCOUNT_PUBLIC], groovy.lang.Closure)
                         uploadToS3.withAWS({role=opensearch-bundle, roleAccount=AWS_ACCOUNT_PUBLIC, duration=900, roleSessionName=jenkins-session}, groovy.lang.Closure)
                            uploadToS3.s3Upload({file=/tmp/src/path, bucket=dummy_bucket, path=/upload/path})
Gradle Test Executor 2 finished executing tests.
Gradle Test Executor 2 finished executing tests.
> Task :test FAILED
83 tests completed, 1 failed, 1 skipped
```

**Comment by peterzhuamazon** [comment]
> test failure
> 
> ```
> TestUploadToS3 > testUploadToS3 STANDARD_OUT
>     Loading shared library jenkins with version <notNeeded>
>     Executing on agent [label:none]
>        UploadToS3_Jenkinsfile.run()
>           UploadToS3_Jenkinsfile.pipeline(groovy.lang.Closure)
>              UploadToS3_Jenkinsfile.echo(Executing on agent [label:none])
>              UploadToS3_Jenkinsfile.stage(sign, groovy.lang.Closure)
>                 UploadToS3_Jenkinsfile.script(groovy.lang.Closure)
>                    UploadToS3_Jenkinsfile.uploadToS3({sourcePath=/tmp/src/path, bucket=dummy_bucket, path=/upload/path})
>                       uploadToS3.string({credentialsId=jenkins-aws-account-public, variable=AWS_ACCOUNT_PUBLIC})
>                       uploadToS3.withCredentials([AWS_ACCOUNT_PUBLIC], groovy.lang.Closure)
>                          uploadToS3.withAWS({role=opensearch-bundle, roleAccount=AWS_ACCOUNT_PUBLIC, duration=900, roleSessionName=jenkins-session}, groovy.lang.Closure)
>                             uploadToS3.s3Upload({file=/tmp/src/path, bucket=dummy_bucket, path=/upload/path})
> Gradle Test Executor 2 finished executing tests.
> Gradle Test Executor 2 finished executing tests.
> > Task :test FAILED
> 83 tests completed, 1 failed, 1 skipped
> ```

Resolved now.

**Review by prudhvigodithi — APPROVED** [review]

File: vars/runGradleCheck.groovy (hunk 2 above)
Comment (owaiskazi19) [reply vars/runGradleCheck.groovy:40]: Can we link the issue here in the comment: https://github.com/opensearch-project/OpenSearch/issues/3674? Later we can run gradle check in the container

File: jenkins/gradle/gradle-check.jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+pipeline {
+    options {
+            timeout(time: 2, unit: 'HOURS')
+    }
+    // gradle check have a lot of issues running on containers
+    // Therefore, we directly run it on the agent node
+    agent {
+        node {
+            // Must use Ubuntu agent with 1 executor or gradle check will show a lot of java-related errors
+            label 'Jenkins-Agent-Ubuntu2004-X64-c54xlarge-Single-Host'
```
Comment (owaiskazi19) [inline jenkins/gradle/gradle-check.jenkinsfile:12]: Can we link the issue here in the comment: https://github.com/opensearch-project/OpenSearch/issues/3674? Later we can run gradle check in the container.

Comment (peterzhuamazon) [reply jenkins/gradle/gradle-check.jenkinsfile:12]: We have decided to not run gradle check in container anymore.
This is also not fully related to container, but to run gradle check on nothing but Ubuntu.
AL2 container will fail and AL2 EC2 will also fail. OS that is not Ubuntu will fail.

File: vars/runGradleCheck.groovy [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+            sh """
+
+                set -e
+                set +x
+
+                env | grep JAVA | grep HOME
+
+                echo "Git clone: ${git_repo_url} with ref: ${git_reference}"
+                rm -rf search
```
Comment (owaiskazi19) [inline vars/runGradleCheck.groovy:26]: Do we need to have `search` directory?

Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:26]: In this instance, yes and it is better to not mix build repo content with opensearch core content.

Comment (owaiskazi19) [reply vars/runGradleCheck.groovy:26]: Build will be placed in a separate directory and rest of the core code will be present separately

Comment (owaiskazi19) [reply vars/runGradleCheck.groovy:26]: The other way I can think of is to have a directory name `opensearch` or `opensearch-core` to be more specific.

File: jenkins/gradle/gradle-check.jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            steps {
+                echo "Install jdk-17"
+                sh("ls ${JAVA17_HOME}/.. | grep jdk-17")
+            }
+        }
+        stage('Run Gradle Check') {
+            steps {
+                script {
+                    sh("ls /var/jenkins/tools/hudson.model.JDK/ && env | grep JAVA")
+                    if (env.USER_BUILD_CAUSE.equals('[]') && env.TIMER_BUILD_CAUSE.equals('[]')) {
```
Comment (owaiskazi19) [inline jenkins/gradle/gradle-check.jenkinsfile:84]: Can we discuss around this check? What purpose does it serves?

File: vars/runGradleCheck.groovy (hunk 5 above)
Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:26]: It is not. Jenkins will pull build repo directly to the workspace without folder.

Comment (peterzhuamazon) [reply vars/runGradleCheck.groovy:26]: As for search folder renaming that can be a follow up improvement PR.

File: jenkins/gradle/gradle-check.jenkinsfile (hunk 6 above)
Comment (peterzhuamazon) [reply jenkins/gradle/gradle-check.jenkinsfile:84]: Just as the variable suggest it is checking what trigger type/cause is it.

**Pull request merged**